content/posts/avl.md

```diff
@@ -0,0 +1,194 @@
+
+# O Problema
+
+A figura abaixo mostra a adição dos elementos [1, 2, 3, 4, 5] em uma ABP.
+
+![abp2](avl/abp-linear.png)
+
+Note que a altura da árvore é 4. Se fossem adicionados $n$ elementos ordenados, a altura seria $n - 1$. Do ponto de vista assintótico, para uma ABP com $n$ nós, a menor altura que uma árvore binária pode ter é $\Theta(log(n))$, enquanto que a maior altura seria $\Theta(n)$. Claramente, quando adicionamos elementos ordenados (de forma crescente ou decrescente) estamos falando do pior caso.
+
+Com a altura árvore $\Theta(n)$ passa a ter a eficiência similar a de uma lista. Por exemplo, a busca em uma árvore com essa altura é $\Theta(n)$. Contudo, sabemos que a árvore binária de pesquisa pode nos fornecer operações $\Theta(log(n))$ se a altura dessa árvore for controlada. Portanto, para garantir a eficiência dos algoritmos é preciso encontrar um esquema que mantenha a árvore com a menor altura possível. Esse esquema é discutido na seção seguinte.
+
+# A Solução: Árvores Balanceadas (AVL)
+
+A solução para resolver o problema apresentado anteriormente é mantermos a árvore balanceada.  Uma árvore pode ser considerada balanceada se a sua altura é $O(log(n))$.
+
+A forma como o balanceamento é feito muda de acordo com o tipo da árvore. Por exemplo, as árvores AVL usam a altura de cada nó para realizar as operações de balanceamento. Já as árvores rubro-negras adicionam propriedades (cores) aos nós e alguns requisitos para mantê-la balanceada.
+
+Neste documento iremos abordar árvores AVL – árvores que utilizam técnicas de balanceamento para manter a altura $O(log(n))$.
+
+O conceito de árvores balanceadas e algoritmos de balanceamento foram introduzidos por **A**delson **V**elskii e **L**andis. Esses dois autores conceberam as árvores AVL, uma árvore **ABP** balanceada. O conceito de árvore balanceada pode ser definido da seguinte maneira:
+
+Uma árvore $<E,raiz,D>$ é balanceada se $|h(E) - h(D)| \le 1$ e se E e D são balanceadas.
+
+Ou seja, as alturas das duas sub-árvores a partir de cada nó diferem no máximo em uma unidade. Dessa maneira, é possível garantir que as operações básicas sejam efetuadas em $O(log(n))$.
+
+**Exemplos de árvores AVL e não-AVL**
+
+![avlnot](avl/avl-not-avl.png)
+
+Provando que $h = O(log(n))$
+
+Número mínimo de nós que em uma AVL de altura h. O pior caso é quando, para cada nó, a sub-árvore à direita tem altura uma unidade maior ou menor do que a sub-árvore à esquerda.
+
+> Mudar as palavras (não entendi exatamente o que quer dizer)
```

Comment:
Cuidado. Isso não é comentário em markdown. Dá uma olhada aqui sobre como comentar: https://stackoverflow.com/questions/4823468/comments-in-markdown